### PR TITLE
Added more support for vertical tables.

### PR DESCRIPTION
**Pull request by Yatskov**

### CODE REVIEW

File: extension/scripts/Selector/SelectorTable.js [hunk 1]
```diff
@@ -106,42 +119,52 @@ var SelectorTable = {
...
+    },
     _getData: function (parentElement) {
 
         var dfd = $.Deferred();
 
         var tables = this.getDataElements(parentElement);
-
+        var verticalTable = this.verticalTable;
+        if (this.verticalTable === undefined) {
+            verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: No need to extract this again here. It should be already in selector object. If it is not then there is a bug in controller.

File: extension/scripts/Selector/SelectorTable.js [hunk 2]
```diff
@@ -106,42 +119,52 @@ var SelectorTable = {
...
-                            }
-                            else {
-                                var header = headerCells[headerCell.header];
+                                var header = headerCells[headerCell.header.replace(/^\s*/,'').replace(/\s*$/,'')];
                                 var rowText = $(dataCell).find(">:nth-child(" + header.index + ")").text().trim();
-                                data[headerCell.name] = rowText;
-                            }
-                        }
+                                if (headerCell.extract){
+                                    data[headerCell.name.replace(/^\s*/,'').replace(/\s*$/,'')] = rowText;
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: Why you need 2 replaces here? Seems that you should already have to this moment processed result keys.

File: extension/scripts/Selector/SelectorTable.js [hunk 3]
```diff
@@ -106,42 +119,52 @@ var SelectorTable = {
...
+
+
 
-            if (isVerticalHeader) {
+            if (verticalTable) {
                 var results = this.getVerticalDataCells(table, dataSelector);
-                result.push.apply(result, results);
-            } else {
+                for (i = 0;i<results.length;i++) {
+                    if (!this.isEmptyObject(results[i])) result.push(results[i]);
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: If you have empty object here. Maybe it is a bug in method getVerticalDataCells

File: extension/scripts/Selector/SelectorTable.js [hunk 4]
```diff
@@ -183,41 +256,69 @@ var SelectorTable = {
...
         }
-    },
 
-    getTableDataRowSelectorFromTableHTML: function (html) {
 
+    },
+    getTableDataRowSelectorFromTableHTML: function (html) {
+        var verticalTable = this.verticalTable;
+        if (this.verticalTable === undefined) {
+            verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: This method should not extract state of selector with jquery.

File: extension/scripts/Selector/SelectorTable.js [hunk 5]
```diff
@@ -153,24 +176,74 @@ var SelectorTable = {
...
+     * Рабочие методы
+     */
+
     getTableHeaderRowSelectorFromTableHTML: function (html) {
         var $table = $(html);
         var firstRow = $table.find("tr:first-child");
         var rowCount = $table.find("tr").length;
+        var verticalTable = this.verticalTable;
+        if (this.verticalTable === undefined) {
+            verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: This method should not extract state of selector with jquery.

File: extension/scripts/Selector/SelectorTable.js [hunk 6]
```diff
@@ -153,24 +176,74 @@ var SelectorTable = {
...
+        var $headerRowColumns = $table.find(headerRowSelector);
+
+        if ($headerRowColumns.length > 1) {
+            $headerRowColumns = $headerRowColumns.find("th:first-child")
+        } else if ($headerRowColumns.find("th").length) {
+            $headerRowColumns = $headerRowColumns.find("th");
+        } else if ($headerRowColumns.find("td").length) {
+            $headerRowColumns = $headerRowColumns.find("td");
+        }
+        verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: This method should not extract state of selector with jquery.

File: extension/scripts/Selector/SelectorTable.js [hunk 7]
```diff
@@ -22,7 +22,7 @@ var SelectorTable = {
         var columns = {};
         var headerRowSelector = this.getTableHeaderRowSelector();
         var $headerRow = $table.find(headerRowSelector);
-        var isVerticalRow = false;
+        var isVerticalRow = this.verticalTable;
 
         if ($headerRow.length > 0) {
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: So many nested if expressions is hard to understand. Whe should discuss how to refactor it.

File: extension/scripts/Selector/SelectorTable.js [hunk 8]
```diff
@@ -106,42 +119,52 @@ var SelectorTable = {
...
+            var dataSelector = this.getTableDataRowSelector();
+
+
 
-            if (isVerticalHeader) {
+            if (verticalTable) {
                 var results = this.getVerticalDataCells(table, dataSelector);
-                result.push.apply(result, results);
-            } else {
+                for (i = 0;i<results.length;i++) {
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: spaces

File: extension/scripts/Selector/SelectorTable.js [hunk 9]
```diff
@@ -153,24 +176,74 @@ var SelectorTable = {
...
+
     getTableHeaderRowSelectorFromTableHTML: function (html) {
         var $table = $(html);
         var firstRow = $table.find("tr:first-child");
         var rowCount = $table.find("tr").length;
+        var verticalTable = this.verticalTable;
+        if (this.verticalTable === undefined) {
+            verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
+        }
         if ($table.find("thead tr:has(td:not(:empty)), thead tr:has(th:not(:empty))").length) {
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: So many nested if expressions is hard to understand. Whe should discuss how to refactor it.

File: extension/scripts/Selector/SelectorTable.js [hunk 10]
```diff
@@ -183,41 +256,69 @@ var SelectorTable = {
...
-    getTableDataRowSelectorFromTableHTML: function (html) {
 
+    },
+    getTableDataRowSelectorFromTableHTML: function (html) {
+        var verticalTable = this.verticalTable;
+        if (this.verticalTable === undefined) {
+            verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
+        }
         var $table = $(html);
         if ($table.find("thead tr:has(td:not(:empty)), thead tr:has(th:not(:empty))").length) {
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: So many nested if expressions is hard to understand. Whe should discuss how to refactor it.

File: extension/scripts/Selector/SelectorTable.js [hunk 11]
```diff
@@ -153,24 +176,74 @@ var SelectorTable = {
...
+            }
+        });
+        }
+        return columns;
     },
 
+
+
+    /**
+     * Рабочие методы
```
Comment (Yatskov) [inline extension/scripts/Selector/SelectorTable.js]: Comments should be in english and describe method and its behaviour.

File: extension/scripts/Selector/SelectorTable.js [hunk 12]
```diff
@@ -64,7 +64,7 @@ var SelectorTable = {
                 if (column.length !== 1) {
                     this.columns.push({
                         header: header,
-                        name: header,
+                        name: name,
```
[the remark this answers is not in the document]
Comment (GooDRomka) [reply extension/scripts/Selector/SelectorTable.js]: I haven't finished with the tables yet and I made this change in the last 15 minutes of work when I was trying to find the place where the error is hiding. I will let you know when I finish the task and ask you to check it out. At the moment, this is a draft in which you are trying to find a mistake, although there should be no logic there.

File: extension/scripts/Controller.js [hunk 13]
```diff
@@ -1190,9 +1192,9 @@ SitemapController.prototype = {
 			if(selector.type === 'SelectorTable') {
 
 				this.getSelectorHTML().done(function(html) {
-
-					var tableHeaderRowSelector = SelectorTable.getTableHeaderRowSelectorFromTableHTML(html);
-					var tableDataRowSelector = SelectorTable.getTableDataRowSelectorFromTableHTML(html);
+					var verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
```
[the remark this answers is not in the document]
Comment (GooDRomka) [reply extension/scripts/Controller.js]: The vertical table is checked every time the Select Selector,SaveSelector,and data preview functions are called. Previously, the controller checked once when we selected the desired type of selector. This is why I do a check before calling the gettableheaderrowselectorfromtablehtml, gettabledatarowselectorfromtablehtml functions

File: extension/scripts/Selector/SelectorTable.js (hunk 2 above)
Comment (GooDRomka) [reply extension/scripts/Selector/SelectorTable.js]: I found this method of removing spaces on a forum with a lot of likes. I consider it necessary to use the working version, and not to invent a Bicycle that may not go.

File: extension/scripts/Selector/SelectorTable.js [hunk 14]
```diff
@@ -158,19 +164,69 @@ var SelectorTable = {
...
+        if ($headerRowColumns.length > 1) {
+            $headerRowColumns = $headerRowColumns.find("th:first-child")
+        } else if ($headerRowColumns.find("th").length) {
+            $headerRowColumns = $headerRowColumns.find("th");
+        } else if ($headerRowColumns.find("td").length) {
+            $headerRowColumns = $headerRowColumns.find("td");
+        }
+        var columns = [];
+        if (verticalTable){
+            $headerRowColumns.prevObject.each(function (i, columnEl) {
```
Comment (mxsnq) [inline extension/scripts/Selector/SelectorTable.js]: - code blocks under if-else alternatives are almost identical, you can do with one
- `if (header.length !== 0)` probably may be replaced with `if (header)`

File: extension/scripts/Selector/SelectorTable.js [hunk 15]
```diff
@@ -158,19 +164,69 @@ var SelectorTable = {
...
+
+    /**
+     * Extract table header column info from html
+     * @param html
+     */
+    getTableHeaderColumnsFromHTML: function (headerRowSelector, html, verticalTable) {
+        var $table = $(html);
+        var $headerRowColumns = $table.find(headerRowSelector);
+
+        if ($headerRowColumns.length > 1) {
```
Comment (mxsnq) [inline extension/scripts/Selector/SelectorTable.js]: Later you use `$headerRowColumns.prevObject` for vertical tables, so I believe this block is intended only for horizontal tables. If so, wrap it under `if (!verticalTable) {...}` condition block and don't use `.prevObject`

File: extension/scripts/Selector/SelectorTable.js [hunk 16]
```diff
@@ -183,41 +239,66 @@ var SelectorTable = {
...
+            if (!verticalTable) {
+                if ($table.find("tr td:not(:empty), tr th:not(:empty)").length) {
+                    var $rows = $table.find("tr");
+                    // first row with data
+                    var rowIndex = $rows.index($rows.filter(":has(td:not(:empty)),:has(th:not(:empty))")[0]);
+                    return "tr:nth-of-type(n+" + (rowIndex + 2) + ")";
+                }
+
+            }
+            if (verticalTable) {
```
Comment (mxsnq) [inline extension/scripts/Selector/SelectorTable.js]: `else`

File: extension/scripts/Selector/SelectorTable.js [hunk 17]
```diff
@@ -183,41 +239,66 @@ var SelectorTable = {
...
+                    var $rows = $table.find("tr");
+                    // first row with data
+                    var rowIndex = $rows.index($rows.filter(":has(td:not(:empty)),:has(th:not(:empty))")[0]);
+                    return "tr:nth-of-type(" + (rowIndex + 1) + ")";
+                } else {
+                    return "";
+                }
+            }
+            else {
+                if (firstRow.find("th").length>0) return "tr>th";
```
Comment (mxsnq) [inline extension/scripts/Selector/SelectorTable.js]: - most coding styles advise to use braces for all control structures (i.e. if, else, etc), even if the body contains only a single statement
- `.length > 0` check may be replaced with `.length` check

File: extension/scripts/Selector/SelectorTable.js [hunk 18]
```diff
@@ -158,19 +164,69 @@ var SelectorTable = {
         });
         return dataColumns;
     },
-
-    getFeatures: function () {
-        return ['selector', 'multiple', 'columns', 'delay', 'tableDataRowSelector', 'tableHeaderRowSelector', 'tableAddMissingColumns']
+    isExtractedDataColumns: function (header) {
+        answer = undefined;
+        this.columns.forEach(function (column) {
```
[the remark this answers is not in the document]
Comment (GooDRomka) [reply extension/scripts/Selector/SelectorTable.js]: Misunderstood this one

[the remark this answers is not in the document]
Comment (mxsnq) [reply extension/scripts/Selector/SelectorTable.js]: You search for element which matches given condition, that's what [`Array.prototype.find`](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/find) is for. You may use it to find the column, then return its name, if its not undefined.

[the remark this answers is not in the document]
Comment (mxsnq) [reply extension/scripts/Selector/SelectorTable.js]: also don't check for `column.extract === true`, just `column.extract`

[the remark this answers is not in the document]
Comment (GooDRomka) [reply extension/scripts/Selector/SelectorTable.js]: This condition have to use  column from this.columns and headerCellName, but in columns.find we can use only column from this.columns, is it?

File: extension/scripts/Controller.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (Yatskov) [reply extension/scripts/Controller.js]: Need @mxsnq opnion here.

File: extension/scripts/Controller.js [hunk 19]
```diff
@@ -1190,9 +1192,9 @@ SitemapController.prototype = {
...
+					var tableHeaderRowSelector = SelectorTable.getTableHeaderRowSelectorFromTableHTML(html,verticalTable);
+					var tableDataRowSelector = SelectorTable.getTableDataRowSelectorFromTableHTML(html,verticalTable);
```
[the remark this answers is not in the document]
Comment (Yatskov) [reply extension/scripts/Controller.js]: No changes still in code style.

File: extension/scripts/Controller.js [hunk 20]
```diff
@@ -877,7 +877,8 @@ SitemapController.prototype = {
 		var tableDataRowSelector = $("#edit-selector [name=tableDataRowSelector]").val();
         var tableHeaderRowSelector = $("#edit-selector [name=tableHeaderRowSelector]").val();
         var tableAddMissingColumns = $("#edit-selector [name=tableAddMissingColumns]").is(":checked");
-		var clickElementSelector = $("#edit-selector [name=clickElementSelector]").val();
+		var verticalTable = $("#edit-selector [name=verticalTable]").is(":checked");
```
[the remark this answers is not in the document]
Comment (Yatskov) [reply extension/scripts/Controller.js]: No changes to the code style here. Still something wrong with tabs and spaces. Why tableHeaderRowSelector, tableAddMissingColumns, clickElementSelector are on another ident.

**Pull request merged**